traffic_ops/app/db/migrations/20200424000000_add_deleted_tables.sql

```diff
@@ -0,0 +1,608 @@
+/*
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+-- +goose Up
+-- SQL in section 'Up' is executed when this migration is applied
+
+CREATE TABLE IF NOT EXISTS deleted_api_capability (
+    id bigserial PRIMARY KEY,
```

Comment:
I've been trying to follow along, but I might've missed something. Why do we need to copy all the existing tables to `deleted_foo`? At a high level, don't we only care about the answer to the question "have any foos been updated or deleted since this point in time?"

Reply:
We need to keep track of the deleted components because, if we just compare the "last_updated" times of the items present in the main tables with what we are supplying in the IMS header, if something got deleted at a time after the IMS time in the request, we would not take that into account and return 304 (which is incorrect, since the endpoint has changed since the IMS time, i.e., one of the objects got deleted).

Reply:
Right, we could only track IDs, but then any change would flag as modified if anything had been deleted, even if it wasn't in the filter. 

E.g. if someone deleted a Parameter on Profile Foo, and someone requested `profiles/name/bar/parameters`, it would flag as modified, even though it wasn't really.

Reply:
Yeah, I understand that part, so we need _something_ to tell us that rows were deleted. But, why do we need _all_ the table data? Shouldn't all the `deleted_foo` tables only have a single column -- e.g. `deleted_at` -- that we update via an `ON DELETE` trigger similar to the trigger for `last_updated`? It seems like the `deleted_foo` tables would only ever need a single row to track when something was last deleted from its corresponding table. Maybe we only need one `deleted` table, which has a row for every _main_ table, that contains the timestamp of when something was last deleted from that referenced table?

Reply:
IMO, it would depend on what all the acceptable params for `DELETE` are. For ex, in some cases, we could be deleting by name, in others, by ID.

Reply:
> Right, we could only track IDs, but then any change would flag as modified if anything had been deleted, even if it wasn't in the filter.

I think there is a trade-off to be made here in terms of complexity vs performance gain. We can keep it really simple by ignoring the "filter" side of things, and have simple invalidations per table. Or, we can copy all of the existing data, which means we have to keep the columns of both tables in sync forever, not to mention dealing with the additional problems created by keeping deleted data forever.

Reply:
Suppose I make a request to `/api/2.0/deliveryservices?cdn=FOO` at 10:00am.
Someone else then does a `DELETE /api/2.0/deliveryservices/cdn-bar-ds` at 10:05am.
I then do an IMS request to `/api/2.0/deliveryservices?cdn=FOO` at 10:10am, with my `If-Modified-Since: 10:00am`.

The `deleted_deliveryservices` table only has `id | last_updated`. It says "there has been a DS deleted since 10:00am, therefore I must send you the full request back.

When in reality, no DS on CDN "FOO" has been deleted or modified, and it can safely return a 304 Not Modified. But there's no way for the code to determine that, without knowing the data queried on, the cdn id, of the deleted record.

We don't actually need all data, just all columns that can be queried on. But that's probably most of them, it seems easier and safer to add them all.

Reply:
In your example, I think it would be fine if we returned the full response back because the reality is that things are not deleted very often, so I don't think the theoretical savings of those rare cases is worth the cost of dev/maintenance. IMO I think it would be better to keep the deletion case simple, not fully duplicate tables, and not fully duplicate queries. If anything was deleted from a table since time T, _any_ IMS requests with time < T that involve those tables could simply return the full response. Since deletion is so rare, the extra complexity doesn't seem worth the savings.

Reply:
>I think there is a trade-off to be made here in terms of complexity vs performance gain. 

I think it's a pretty big performance gain, and not too complex. Without it, it'll mean any modification to any profile will make IMS fail on ORT for every server. It's hard to know for sure without looking at detailed change data over a long period of time, but I suspect not handling params will put the IMS hit rate <10%, and with params will be >90%.

> we have to keep the columns of both tables in sync forever

We do. It's a little code, but it doesn't seem unreasonable, it just means adding a column has to add it in both tables.

> not to mention dealing with the additional problems created by keeping deleted data forever.

Personally, I don't believe this will be an issue. Our data is tiny. Speaking from experience, gigabytes of data really isn't big or hard to deal with, and our data is megabytes.

But if it is an issue, it's pretty easy to truncate it, and always return Modified for an IMS older than that.

Reply:
>I don't think the theoretical savings of those rare cases is worth the cost of dev/maintenance. 

I guess we'll have to agree to disagree on that. The savings are not theoretical, and I don't believe the cases are rare. I believe the performance savings are huge. For ORT, TP, and everything else that regularly hits TO.

And I believe the development and maintenance cost are very small.

>fully duplicate tables

I don't see why this is such a big deal. "Duplicate tables" are just a few lines of code. It's a `create table deleted_foo (id, bar, baz...`, and then for migrations,
```
ALTER TABLE deliveryservice ADD COLUMN ecs_enabled boolean;
ALTER TABLE deleted_deliveryservice ADD COLUMN ecs_enabled boolean;
```

It's a very small amount of code, and a very small amount of data. 

Yes, it's a maintenance cost, like every line of code. But why is it large, or difficult, or dangerous, more than any other dozen of code?

>fully duplicate queries

The query is not "fully duplicated" in code, the query parameters are abstracted, so it's

```
SELECT max(t) from (
SELECT max(last_updated) as t from `+ tableName + ` q ` + where + orderBy + pagination +
` UNION ALL
select max(last_updated) as t from deleted_`+ tableName + ` q ` + where + orderBy + pagination +
` ) as res`
```

Rather than 
```
SELECT max(t) from (
SELECT max(last_updated) as t from `+ tableName + ` q ` + where + orderBy + pagination +
` UNION ALL
select max(last_updated) as t from deleted_`+ tableName + ` where id = $1`
` ) as res`
```

The query difference is pretty small.

Reply:
>  Without it, it'll mean any modification to any profile will make IMS fail on ORT for every server.

But we have a plan to address this in the near future (cache config snapshots), to where this would no longer matter for ORT anyways. So really, the IMS here is really for optimizing things like SREs doing things in Traffic Portal, etc, not for ORT (long-term).

> it doesn't seem unreasonable, it just means adding a column has to add it in both tables

That is where we get into trouble. We already have problems with keeping our nullable and non-nullable Go structs in sync with each other. Unless it's been hardcoded into your brain that you need to update things in all the various places that they're duplicated, you will forget to do so. Duplication is not ideal. It's also an additional, ongoing cost to developing new features, not just maintenance.

>  Our data is tiny.

I'm not really concerned about the size of the data, I'm more concerned about the _contents_ of the data. Data is a liability. We have things like passwords, physical addresses, IP addresses, etc. -- all things that don't really make me comfortable knowing that we'd be holding onto them indefinitely.

> The savings are not theoretical

By theoretical I meant in the sense of the _amount_ of savings. I'm not disagreeing that successful IMS requests are _actual_ savings.

> and I don't believe the cases are rare.

I guess that really depends on what we each think "rare" means. What do you consider "rare"? Once a day? A week? A month?

For reference, in the last ninety days at my company, we only issued 104 `DELETE` TO API requests. Now, that isn't including requests like `POST /profileparameter` that actually delete things under the hood, but we had 12 of those `POST /profileparameter` requests in the last ninety days. So we're talking about slightly over one thing a day on average that could potentially invalidate cache for a certain table. Also, they appeared to mostly be DELETEs of servers and delivery services, which make sense, but it's important to also note that those are also the kinds of resources that are typically asked for in large quantities via GET requests.

Think about opening either of those tables (servers and DSes) in TP -- by default TP requests _all_ of those, and those are easily the two slowest pages in TP. If there were _any_ servers or DSes deleted, it doesn't matter if we tracked every single one deleted or simply tracked that certain tables had been deleted from. That cache would be invalidated the same either way. Why not try to keep the implementation simple? 

> The query is not "fully duplicated" in code, the query parameters are abstracted

I was referring to the `insert into deleted_table` queries -- they contain a lot of duplicated code with their corresponding `insert into table` and `select from table` queries. We would need to be careful to share as much as possible between those three queries rather than having three separate, mostly-duplicated queries, to keep in sync.

Reply:
>we have a plan to address this in the near future (cache config snapshots)

I see solutions like denormalized outside-the-API JSON blob snapshots or intermediary proxies as hacks to work around a broken system, not the ideal. If Traffic Ops were reasonably performant and implemented HTTP standards, and relatively simple performance optimizations like IMS, a 1s cache, and Read-While-Writer on concurrent requests, expensive and unsafe workarounds wouldn't be necessary. 

There are scales where the cost of these things becomes necessary; <1 req/s on a <100MB database are several orders of magnitude below those levels.

>in the last ninety days at my company, we only issued 104 DELETE

Even one DELETE a day, if it's breaking an otherwise successful IMS, means increasing the TO load by thousands of times the necessary, for some brief period. If that change was one cache, it means the server has to handle full requests from every cache in the CDN checking in at once, instead of a single full request and 999 cheap IMS's.

>We would need to be careful to share as much as possible between those three queries rather than having three separate, mostly-duplicated queries

String building isn't hard. In fact, we already have columns duplicated in places, which should really concatenate strings to avoid duplicating all the column names.

At any rate, I'm not going to -1. I'm not a primary developer on Traffic Ops anymore, and I really don't like telling other people how hard something should be for them. If decision is that it's too difficult and expensive to maintain real IMS, well, I'm not writing (much) code on the project anymore to say.

But I will say, Traffic Ops is at the center of everything else in Traffic Control. If TO would do the right thing -- HTTP standards like IMS and Delta, Read-While-Writer, 1s caching, other small things to get decent performance -- these things would be a tiny, tiny fraction of the code TO has to maintain, and would make things better for every other component of Traffic Control, as well as all its operators and users.

The cost of these small features in TO is less than cached blobs and slower configs, and far, far less than additional intermediary proxy service deployments.

TO has had abysmal performance since the beginning. Golang is much better now, but the scalability of TO, how many requests it can handle, is still orders of magnitude lower than it should be.

The things we have to do to work around poor TO performance -- proxies in front of TO, ORT dispersion and backoff and slower config deployment, TM slower health updating -- everything else in TC would be so, so much better if we were just willing to do the work to make TO performant. And the development and operational time of all those things far, far outweighs the time for TO.

Reply:
Looking around at the way other projects implement this, it looks like they typically do one of two things:

- Implement a `last_deleted` timestamp table with one row for each table that just stores the timestamp of the last deletion. That returns payloads which invalidate unnecessarily sometimes but are still a useful improvement over nothing, or
- Don't support IMS on collections of objects. Just use ETags for collections, and support IMS on single objects.

Those both seem fine to me - and I didn't look too hard at the DB implementation tbh because I thought he wanted me specifically to look at the API code and it's still just a draft - but I share Rawlin's reservations about creating tables to shadow all data that ever existed in the database.

Reply:
Refactored the DELETE logic to use just one table with just the last timestamp of the deleted member of that particular "type".

Reply:
>  I see solutions like denormalized outside-the-API JSON blob snapshots or intermediary proxies as hacks to work around a broken system, not the ideal.

I wouldn't really call it a hack. I think it's actually a step in the right direction that will help us achieve much greater performance and efficiency in dishing out config data to thousands of caches at once. Since caches are by far the biggest client base of the TO API, it makes sense to optimize their requests as much as possible, and I think the cache-config snapshots idea is the simplest way to give us the most optimal solution for those requests. It can even have all those nice features -- perfect IMS, delta encoding, read-while-writer, 1s caching -- at a much cheaper price than implementing them all across the entire API since it's just a single endpoint.

_Back to the PR at hand:_

I've also been thinking, we really need to make sure the IMS DB queries are significantly cheaper compared to the non-IMS DB queries. Currently, I'm pretty sure that they will do a full sequential scan in order to find the most recent `last_updated` because it isn't indexed, and that isn't really much cheaper on the DB than sequentially scanning and returning all of the data. So IMS hits would save bandwidth, but not load on the DB, which I think may be more important. It also basically doubles the cost of an IMS miss, and we don't really know what our IMS hit percentage would be in practice.

I was playing around with indexing `last_updated`, and this seems to make the `select max(last_updated)` queries much cheaper:
```
create index deliveryservice_last_updated_idx on deliveryservice (last_updated DESC NULLS LAST);
```
The `DESC` seems to be important because, in theory, the first timestamp in the index should be the max, so this should be a very cheap query compared to a sequential scan.

There's also the chance that we're wrong about the DB performance in practice, if things are updated often enough to cause way more IMS misses than expected. In that case, it would be nice to have a feature flag to disable IMS handling. That also means we should probably log whether a request was IMS_HIT, IMS_MISS, NON_IMS, etc. Then we will be able to determine what our hit percentage is from the logs.

Reply:
Ah, yes, if `last_updated` doesn't have an index, it definitely needs one for these queries. +1
And +1 on logging